### PR TITLE
Add devise

### PR DESCRIPTION
**Pull request by rica213**
Hi, 
I am Larissa
I have collaborated with Mert (@mertkantaroglu ) in this milestone.
The followings are what we have done:
- Remove the current_user method from the application controller
- Install and set up the devise gem
- Update readme file

Happy reviewing!

### CODE REVIEW

File: db/migrate/20230531092621_add_confirmation_fields_to_users.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class AddConfirmationFieldsToUsers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :users, :confirmation_token, :string
+    add_column :users, :confirmed_at, :datetime
+    add_column :users, :confirmation_sent_at, :datetime
+    add_column :users, :unconfirmed_email, :string
+  end
+end
```
Comment (TedLivist) [inline db/migrate/20230531092621_add_confirmation_fields_to_users.rb:8]: - [ ] Some of the migrations here are preventing me from setting up your app. You have some columns that you are trying to create which already exist in a previous migration. You may need to delete this migration or only keep the columns that don't exist previously.

  image

**Review by TedLivist — COMMENTED** [review]
## STATUS: INVALID CODE REVIEW REQUEST 🛑 🛑

Hi @rica213 and @mertkantaroglu,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

### I am unable to set up your app due to migration issues. Kindly fix them, the issue is highlighted below.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: db/migrate/20230531092621_add_confirmation_fields_to_users.rb (hunk 1 above)
Comment (rica213) [reply db/migrate/20230531092621_add_confirmation_fields_to_users.rb:8]: Hi @TedLivist ,
Thank you for your time. Indeed we had this problem when working on my partner's machine, it only works on mine. We tried to fix it but nothing. Would you give us some guidance on how to fix that issue? Thanks in advance
Larissa

Comment (TedLivist) [reply db/migrate/20230531092621_add_confirmation_fields_to_users.rb]: Alright @rica213 I see you have deleted the file as suggested. That's the solution.

File: app/models/user.rb [hunk 2]
```diff
@@ -1,4 +1,8 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (TedLivist) [inline app/models/user.rb:5]: - [ ] Users should have to confirm their email after signing up. Kindly add the `:confirmable` module to the configuration.

File: app/views/devise/registrations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  <div class="field">
+    <%= f.label :password_confirmation %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Sign up" %>
+  </div>
+<% end %>
```
Comment (TedLivist) [inline app/views/devise/registrations/new.html.erb:27]: - [ ] Kindly add the field for `name` during sign-up as there is a validation error as a result of its absence and I cannot sign up a user.

- [ ] Kindly center your form on the page as we have on the wireframe layout.

  image

File: app/views/devise/sessions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      <%= f.label :remember_me %>
+    </div>
+  <% end %>
+
+  <div class="actions">
+    <%= f.submit "Log in" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (TedLivist) [inline app/views/devise/sessions/new.html.erb:26]: - [ ] Kindly center the login form and fix your layout for the check box and its label.

  image

File: app/views/devise/sessions/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+  <% if devise_mapping.rememberable? %>
+    <div class="field">
+      <%= f.check_box :remember_me %>
+      <%= f.label :remember_me %>
+    </div>
+  <% end %>
+
+  <div class="actions">
+    <%= f.submit "Log in" %>
```
Comment (TedLivist) [inline app/views/devise/sessions/new.html.erb:22]: - [ ] **[OPTIONAL]** I create a user from the console and decided to log but I get this error page after signing in. Kindly fix that.

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -1,6 +1,2 @@
 class ApplicationController < ActionController::Base
-  # Make current user data available to all controllers
```
Comment (TedLivist) [inline app/controllers/application_controller.rb:2]: - [ ] We should not allow non-logged-in users to access different pages of our app. Kindly add the `authenticate_user!` helper to the application controller.

  ```
  before_action :authenticate_user!
  ```

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @rica213 and @mertkantaroglu,

Good job so far!
✔️ Integration of devise
✔️ Users can change their forgotten passwords
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ichala — APPROVED** [review]
Hi Student  ,
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:  
 
68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f6c3366516333614356304871707a5763552f67697068792e676966

# To Highlight 💪🏻

- Changes Implemented ✔️ 


 Good job implementing the requirements  :+1: Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Optional Suggestions  :

- N/A


**********
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag  ichala in your question so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**